userspace/libscap/engine/gvisor/gvisor.h

```diff
@@ -74,4 +79,33 @@ typedef struct parse_result parse_result;
 parse_result parse_gvisor_proto(scap_const_sized_buffer gvisor_buf, scap_sized_buffer scap_buf);
 
 } // namespace parsers
+
+
+class engine {
+public:
+    engine(char *lasterr);
+    ~engine();
+    int32_t init(std::string socket_path);
+    int32_t close();
+
+    int32_t start_capture();
+    int32_t stop_capture();
+
+    int32_t next(scap_evt **pevent, uint16_t *pcpuid);
+    
+private:
+    int32_t process_message_from_fd(int fd);
+    void free_sandbox_buffers();
+
+    char *m_lasterr;
+    int m_listenfd;
+    int m_epollfd;
+    std::string m_socket_path;
+    std::thread m_accept_thread;
+    std::deque<scap_evt *> m_event_queue{};
```

Comment:
Who owns the memory pointed to the elements in the queue? When is it allocated/freed?

Reply:
The memory always comes from `m_sandbox_buffers`. The pointers saved in the queue is always within those buffers, which can only be modified/resized/overwritten if the queue is empty.